crates/containerd-shim-wasm/src/sandbox/cgroups.rs

```diff
@@ -0,0 +1,674 @@
+            if let Some(swap) = memory.swap() {
+                ensure_write_file(
+                    controller_path.join("memory.memsw.limit_in_bytes"),
+                    &swap.to_string(),
+                )?;
+            } else {
+                // If memory is unlimited and swap is not explicitly set, set swap to unlimited
+                // See https://github.com/opencontainers/runc/blob/eddf35e5462e2a9f24d8279874a84cfc8b8453c2/libcontainer/cgroups/fs/memory.go#L70-L71
+                if mem_unlimited {
+                    ensure_write_file(
+                        controller_path.join("memory.memsw.limit_in_bytes"),
+                        &"-1".to_string(),
+                    )?;
+                }
+            }
+            if let Some(reservation) = memory.reservation() {
+                ensure_write_file(
+                    controller_path.join("memory.soft_limit_in_bytes"),
+                    &reservation.to_string(),
+                )?;
+            }
+            if let Some(swappiness) = memory.swappiness() {
+                ensure_write_file(
+                    controller_path.join("memory.swappiness"),
+                    &swappiness.to_string(),
+                )?;
+            }
+            if let Some(kernel) = memory.kernel() {
+                ensure_write_file(
+                    controller_path.join("memory.kmem.limit_in_bytes"),
+                    &kernel.to_string(),
+                )?;
+            }
+            if let Some(kernel_tcp) = memory.kernel_tcp() {
+                ensure_write_file(
+                    controller_path.join("memory.kmem.tcp.limit_in_bytes"),
+                    &kernel_tcp.to_string(),
+                )?;
+            }
+            if let Some(oom_kill_disable) = memory.disable_oom_killer() {
+                if oom_kill_disable {
+                    ensure_write_file(controller_path.join("memory.oom_control"), "1")?;
+                }
+            }
+        }
+
+        if let Some(pids) = &res.pids() {
+            let controller_path = self.get_controller("pids")?;
+            ensure_write_file(controller_path.join("pids.max"), &pids.limit().to_string())?;
+        }
+
+        if let Some(hugepages) = res.hugepage_limits() {
+            for page in hugepages {
+                let controller_path = self.get_controller("hugetlb")?;
+                let path =
+                    controller_path.join(format!("hugetlb.{}.limit_in_bytes", page.page_size()));
+                ensure_write_file(path, &page.limit().to_string())?;
+            }
+        }
+
+        if let Some(blkio) = res.block_io() {
+            let controller_path = self.get_controller("blkio")?;
+            if let Some(weight) = blkio.weight() {
+                ensure_write_file(controller_path.join("blkio.weight"), &weight.to_string())?;
+            }
+            if let Some(weight_device) = blkio.weight_device() {
+                let mut file = OpenOptions::new()
```

Comment:
Indeed I considered this, but went with this since I'm writing to the same file multiple times.
But I can definitely switch it to `ensure_write_file` to simplify things. Trade-off here is is overhead opening/closing repeatedly vs 1 open/close for the entire thing.